cpp/include/raft/neighbors/detail/knn_brute_force.cuh

```diff
@@ -0,0 +1,588 @@
+                                                    true,
+                                                    metric_arg);
+      if (metric == raft::distance::DistanceType::L2Expanded ||
+          metric == raft::distance::DistanceType::L2SqrtExpanded) {
+        auto row_norms = search_norms.data() + i;
+        auto col_norms = index_norms.data() + j;
+        auto dist      = temp_distances.data();
+        auto count     = thrust::make_counting_iterator<IndexType>(0);
+
+        thrust::for_each(handle.get_thrust_policy(),
+                         count,
+                         count + current_query_size * current_centroid_size,
+                         [=] __device__(IndexType i) {
+                           IndexType row = i / current_centroid_size,
+                                     col = i % current_centroid_size;
+
+                           auto val = row_norms[row] + col_norms[col] - 2.0 * dist[i];
+
+                           // due to numerical instability (especially around self-distance)
+                           // the distances here could be slightly negative, which will
+                           // cause NaN values in the subsequent sqrt. Clamp to 0
+                           dist[i] = val * (val >= 0.0001);
+                           if (metric == raft::distance::DistanceType::L2SqrtExpanded) {
+                             dist[i] = sqrt(dist[i]);
+                           }
+                         });
+      }
+
+      detail::select_k<IndexType, ElementType>(temp_distances.data(),
+                                               nullptr,
+                                               current_query_size,
+                                               current_centroid_size,
+                                               distances + i * k,
+                                               indices + i * k,
+                                               select_min,
+                                               current_k,
+                                               stream);
+
+      // if we're tiling over columns, we need to do a couple things to fix up
+      // the output of select_k
+      // 1. The column id's in the output are relative to the tile, so we need
+      // to adjust the column ids by adding the column the tile starts at (j)
+      // 2. select_k writes out output in a row-major format, which means we
+      // can't just concat the output of all the tiles and do a select_k on the
+      // concatenation.
+      // Fix both of these problems in a single pass here
+      if (tile_cols != n) {
+        const ElementType* in_distances = distances + i * k;
+        const IndexType* in_indices     = indices + i * k;
+        ElementType* out_distances      = temp_out_distances.data();
+        IndexType* out_indices          = temp_out_indices.data();
+
+        auto count = thrust::make_counting_iterator<IndexType>(0);
+        thrust::for_each(handle.get_thrust_policy(),
```

Comment:
from our conversation in slack - we're going to leave this one as-is for now. The issue is that map_offset isn't a drop in replacement for for_each : in this case we're writing two outputs instead of one, and calculating the output index from the input index